cmd/clusterctl/client/repository/repository_github.go

```diff
@@ -43,7 +43,8 @@ import (
 
 const (
 	httpsScheme                    = "https"
-	githubDomain                   = "github.com"
+	publicGithubDomain             = "github.com"
+	githubHostPrefix               = "github."
```

Comment:
I think its not enough to clarify "it is github" via "the domain is `github.*`.

A private github instance could also be at `git.foocompany.country`.

But the same applies for a private gitlab instance.

In general, I'm fine with this approach. But maybe there is a better way for private instances, e.g. use a sepcial scheme for setting it up. E.g. have:

* `github://git.foocompany.country` which then uses the github client implementation (not httpGet which is also in `repository_github.go`) (only allowing TLS)
* `gitlab://git.foocompany.country` which then uses the gitlab implementation (only allowing TLS)

And to not break existing use cases, keep:
* `https://github.com` to use the httpGet method and fallback to the github client
* `https://gitlab.*` to use the gitlab implementation

But before jumping into an implementation, I'm happy to hear thoughts of Maintainers about this :-)

Reply:
I agree that this is not a perfect implementation to identify github vs gitlab repositories using the domain. It was based on the existing techniques used to identify private gitlab domains in the repo. I like the idea of using httpScheme and I'm happy to pivot in that direction.

While other maintainers weigh in, what's the best place to continue this discussion? Is it this PR / the linked issue or over at cluster-api slack? I think this is a feature that cluster-api should support going forward.

Reply:
I agree with @chrischdi in that this might not cover all cases. It seems like GitHub pages wouldn't work either since the domain name could start with the username like `http://jont828.github.io`. However, I do think it's reasonable to go with something like @chrischdi's suggestion in a crawl-walk-run iteration (and make sure we don't break existing use cases).